Add GET /safes/<safe-address>/multisig-transactions route

hectorgomezv opened the pull request:
Closes #159 

Add GET /safes/<safe-address>/multisig-transactions route

- [X] Caching support
- [X] Unit/controller tests
- [X] End-to-end tests
- [X] OpenApi specification

hectorgomezv commented on src/routes/transactions/transactions.service.ts:
```diff
@@ -0,0 +1,59 @@
+    executed?: boolean,
+    limit?: number,
+    offset?: number,
+  ): Promise<Partial<Page<MultisigTransaction>>> {
```
This uses `Partial<T>` to comply with the current implementation, which does not return the `count` attribute in this route.

fmrsabino commented on src/routes/transactions/transactions.module.ts:
```diff
@@ -0,0 +1,15 @@
+@Module({
+  controllers: [TransactionsController],
+  providers: [
+    AddressInfoHelper,
```
Here the `AddressInfoModule` should be imported

hectorgomezv replied:
Changed in [6d58c66](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/6d58c665e46105677aa5f746b9ac461b4e0c8ee3)

fmrsabino commented on src/routes/transactions/transactions.service.ts:
```diff
@@ -0,0 +1,59 @@
+
+    const safeInfo = await this.safeRepository.getSafe(chainId, safeAddress);
+    const results = await Promise.all(
+      transactions.results.map(async (multiSignTransaction) => ({
```
typo: `multiSignTransaction`

hectorgomezv replied:
Changed in [6d58c66](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/6d58c665e46105677aa5f746b9ac461b4e0c8ee3)

fmrsabino commented on src/routes/transactions/transactions.service.ts:
```diff
@@ -0,0 +1,59 @@
+    );
+
+    return {
+      next: transactions.next, // TODO: map URL
+      previous: transactions.previous, // TODO: map URL
+      results,
```
We have the tools in place to map this easily.

You can check the `RouteUrlDecorator`, `PaginationDataDecorator` and the functions under `pagination.data.ts`

hectorgomezv replied:
Changed in [6d58c66](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/6d58c665e46105677aa5f746b9ac461b4e0c8ee3)

fmrsabino commented on src/domain/tokens/entities/schemas/token.schema.ts:
```diff
@@ -5,11 +5,11 @@ export const tokenSchema: JSONSchemaType<Token> = {
   properties: {
     address: { type: 'string' },
-    decimals: { type: 'number' },
+    decimals: { type: 'number', nullable: true },
```
Good catch. This only applies to `ERC721` but maybe we should split this schema into multiple ones and use a `discriminator` for the `type`.

Something that can be tackled separately. 👍

fmrsabino commented on src/domain/tokens/__tests__/token.factory.ts:
```diff
@@ -0,0 +1,20 @@
+
+export default function (
+  address?: string,
+  decimals?: number,
```
If `decimals` is null we should allow that here too.

hectorgomezv replied:
Changed in [6d58c66](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/6d58c665e46105677aa5f746b9ac461b4e0c8ee3)

fmrsabino commented on src/routes/transactions/entities/multisig-transaction.entity.ts:
```diff
@@ -0,0 +1,30 @@
+  type: string;
+}
+
+export class TransactionSummary {
```
OOC: why don't we have Swagger decorators on some of the classes on this route?

fmrsabino commented on src/routes/transactions/entities/transfer-transaction.entity.ts:
```diff
@@ -0,0 +1,35 @@
+}
+
+export interface TransferTransaction extends TransactionInfo {
+  sender: unknown;
+  recipient: unknown;
```
Why are these `unkown`? Shouldn't they be `AddressInfo`?

fmrsabino reviewed (changes requested):
Still need to review some changes. This is more for early feedback/clarifications.

usame-algan commented on src/routes/transactions/transactions.controller.spec.ts:
```diff
@@ -0,0 +1,271 @@
+    app = moduleFixture.createNestApplication();
+    app.useGlobalFilters(new DataSourceErrorFilter());
+
+    await app.init();
```
Do we need to init the app before running each test or could it work with only doing it once? When I move this part into `beforeAll`, some tests randomly fail with a 503 error. In the [nestjs documentation](https://docs.nestjs.com/fundamentals/testing) they are also calling `init` in `beforeAll`.

hectorgomezv replied:
Yes, good catch, I also tried to init the app once in the `beforeAll` hook, and noted that. I'm not sure why this is happening, maybe it's something regarding the app configuration. I didn't spend so much time on it honestly but could be interesting to tackle it in a separate issue since it's not specific to this PR. Thanks!

usame-algan commented on src/routes/transactions/transactions.controller.spec.ts:
```diff
@@ -0,0 +1,271 @@
+        data: { results: ['invalidData'] },
+      });
+
+      await request(app.getHttpServer())
```
Do we need to await here if there is nothing after this call?

hectorgomezv replied:
I tried it in [9590b4e](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/9590b4edc09b406d88e253fcdf98e6725d60da84) and didn't work. I think not awaiting the Promise to resolve is letting Jest execute the `afterAll` hook and therefore closing down the app, I think that's the reason I was getting a 503 when removing the `await` instruction 🙂

moisses89 commented on src/routes/transactions/entities/multisig-transaction.entity.ts:
```diff
@@ -0,0 +1,41 @@
+  @ApiProperty()
+  confirmationsSubmitted: number;
+  @ApiPropertyOptional()
+  missingSigners?: string[];
```
`missingSigners` are not really an `array` of strings, https://github.com/safe-global/safe-client-gateway/blob/main/src/routes/transactions/models/summary.rs#L155

hectorgomezv replied:
Changed in [85720bd](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/85720bdf33a94eaab2d886ffe53d1e0f35df0059)

moisses89 commented on src/routes/transactions/entities/multisig-transaction.entity.ts:
```diff
@@ -0,0 +1,41 @@
+  @ApiProperty()
+  txInfo: TransactionInfo;
+  @ApiProperty()
+  executionInfo: ExecutionInfo;
```
I think that this is optional and there is other one that is `SafeAppInfo`.

hectorgomezv replied:
True, I forgot to include it 😟 maybe because `SafeAppInfo` wasn't included in any of the transactions I used to test the functionality. But I think this deserves special attention because it seems there's some retrieval logic (see [safe_app_info.rs](https://github.com/safe-global/safe-client-gateway/blob/main/src/routes/transactions/converters/safe_app_info.rs)) and I'm not sure we need to change something in the domain because `origin` seems to be a JSON field. 

I included the entities in [ea6fec6](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/ea6fec62a715b307096688b11b1a4da1b39de46f) and I think it's better to let this as a TODO and tackle it in a separate PR, wdyt? (Any help from @fmrsabino is really appreciated since he probably has more context on this)

moisses89 replied:
I would like to include it in this PR because is part of the `multisigtransaction` endpoint, if we forget include it we could break things on the client side. 
I was searching an example but while the tx-service returns the `origin`, the client-gateway is not including the `safe_app_info`.

hectorgomezv replied:
Ok, I was suggesting it to avoid doing too many things in this PR, it is complex enough I think. Closing this PR doesn't mean this feature/route is ready for production if we create an additional issue to tackle it. But I'm fine doing it here if you prefer 🙂 I'd like to have some inputs from @fmrsabino regarding this if possible before implementing the `origin` mapping.

hectorgomezv replied:
Done 🙂 https://github.com/5afe/safe-client-gateway-nest/issues/201

moisses89 reviewed (changes requested):
Good work!
I think that is a very hard schema with a lot of entities related, then is normal make mistakes :)

fmrsabino commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,592 @@
+    const recipient =
+      transaction.safe !== transaction.to
+        ? await this.addressInfoHelper.getOrDefault(chainId, transaction.to)
+        : { value: transaction.to };
```
Why do we have this fallback?

hectorgomezv replied:
I just rewrote it from the Rust implementation https://github.com/safe-global/safe-client-gateway/blob/8855fec20cb1ac2cb9783802dad35ba1071578bc/src/common/converters/mod.rs#L25

Maybe I'm missing something and the fallback is not necessary, do you think we could just use `transaction.to` value directly here?

fmrsabino replied:
I think this is related with some other comments that I made: https://github.com/5afe/safe-client-gateway-nest/pull/197#discussion_r1045949407 and https://github.com/5afe/safe-client-gateway-nest/pull/197#discussion_r1045599136

We should just return here whatever the `addressInfoHelper` returns. Calling `addressInfoHelper.getOrDefault` and then having another default seems a bit strange 🤔

hectorgomezv replied:
I was trying to mimic the exact response from the Rust CGW implementation, even for `nulls`, that's the reason for that defaults, but I agree maybe it doesn't make much sense. I removed all the fallbacks to rely on `AddressInfoHelper`. It's changed in [a70656a](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/a70656aa38e4762de4339cba81c6ec7ff0d4fd13)

fmrsabino reviewed (changes requested):
Some questions. Focused more on the mapper functionality. Still reviewing other parts.

hectorgomezv commented on src/datasources/transaction-api/transaction-api.service.ts:
```diff
@@ -293,21 +293,31 @@ export class TransactionApi implements ITransactionApi {
+    executionDateLte?: string,
+    to?: string,
+    value?: string,
+    nonce?: string,
```
Hmm, not sure what you mean, do you mean we should convert them to `number` internally? The input type is `string` (see https://safe-client.safe.global/index.html) and those values are sent using a query param to the transaction service. In fact, I just tried to use them for filtering and it works as expected 🤔

hectorgomezv commented on src/routes/transactions/__tests__/get-multisig-transactions.e2e-spec.ts:
```diff
@@ -0,0 +1,123 @@
+      .expect(200)
+      .then(({ body }) => {
+        expect(sortTransactionsByTimestamp(body)).toEqual(
+          sortTransactionsByTimestamp(expectedResponse),
```
Yes, it should, I experienced divergences because of different `modified` timestamps between Goerli staging and production services, but it shouldn't be necessary anymore 👍 I'll remove the sorting.

hectorgomezv replied:
Changed in [85720bd](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/85720bdf33a94eaab2d886ffe53d1e0f35df0059)

hectorgomezv commented on src/domain/interfaces/transaction-api.interface.ts:
```diff
@@ -100,6 +100,11 @@ export interface ITransactionApi {
+    to?: string,
+    value?: string,
+    nonce?: string,
     limit?: number,
```
Changed in [85720bd](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/85720bdf33a94eaab2d886ffe53d1e0f35df0059)

hectorgomezv commented on src/routes/transactions/entities/custom-transaction.entity.ts:
```diff
@@ -0,0 +1,18 @@
+  @ApiProperty()
+  value: string;
+  @ApiProperty()
+  methodName: string;
```
Changed in [85720bd](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/85720bdf33a94eaab2d886ffe53d1e0f35df0059)

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,592 @@
+    transaction: MultisigTransaction,
+    safe: Safe,
+  ): number {
+    return transaction.confirmationsRequired ?? safe.threshold;
```
Good question. Maybe when the transaction was created the number of required signers is recorded in the `confirmationsRequired` field, and It could be different to `safe.threshold` because the Safe configuration could change after the transaction was created. Is that possible?

fmrsabino replied:
With caching in place for both endpoints it's hard to say. I'd say to leave a `TODO` here so we can come back to this later.

But It'd be better to rely only on one source.

hectorgomezv replied:
Added a TODO in [4827c02](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/4827c02120e9672707846c48d0552b30dd1952cb)

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,592 @@
+    }
+  }
+
+  private mapTxStatus(transaction: MultisigTransaction, safe: Safe): string {
```
Changed in [20294e3](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/20294e3902127f7a18dd407d5d72ad9f5cd97e86)

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,592 @@
+  private readonly CHANGE_MASTER_COPY = 'changeMasterCopy';
+  private readonly ENABLE_MODULE = 'enableModule';
+  private readonly DISABLE_MODULE = 'disableModule';
+  private readonly SET_GUARD = 'setGuard';
```
Changed in [20294e3](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/20294e3902127f7a18dd407d5d72ad9f5cd97e86)

hectorgomezv commented on src/domain/tokens/__tests__/token.factory.ts:
```diff
@@ -0,0 +1,20 @@
+    logoUri: logoUri ?? faker.internet.url(),
+    name: name ?? faker.random.word(),
+    symbol: symbol ?? faker.finance.currencySymbol(),
+    type: type ?? TokenType.Erc20,
```
Changed in [20294e3](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/20294e3902127f7a18dd407d5d72ad9f5cd97e86)

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,592 @@
+      transaction.to,
+    );
+    return {
+      type: 'Custom',
```
Do you mean extracting the multisig transaction types to an enum? Or using a constructor/builder with a fixed property value for `type`? Could you elaborate, please? :)

fmrsabino replied:
You return `CustomTransactionInfo`. Does `CustomTransactionInfo` have `Custom` always as the value of `type`? If yes we don't need to parameterise it no?

hectorgomezv replied:
That's right 👍🏻 But `CustomTransactionInfo` is a JS class, if we want to have a fixed value for one of its properties we'd need a constructor. Maybe this could be a good candidate to use a builder pattern, wdyt?

fmrsabino replied:
Yes I think a constructor would be a good option here:

```typescript
export class CustomTransactionInfo extends TransactionInfo {
  @ApiProperty()
  type = 'Custom';

  ...

  constructor(
    to: AddressInfo | null,
    dataSize: string,
    value: string,
    methodName: string,
    actionCount: number | null,
    isCancellation: boolean,
  ) {
    super();
    this.to = to;
    this.dataSize = dataSize;
    this.value = value;
    this.methodName = methodName;
    this.actionCount = actionCount;
    this.isCancellation = isCancellation;
  }
}
```

Also I think that instead of having `?` in the route properties we should have `T | null`. At this point we should know if the value is present or not and I think it's more semantically correct. Wdyt?

hectorgomezv replied:
I agree, changed in [de4181b](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/de4181bab770581f3e869e7eadce9fd43d16e5a9)

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,592 @@
+  ): Promise<TransactionInfo> {
+    const value = Number(transaction?.value) || 0;
+    const dataSize = transaction.data
+      ? (Buffer.byteLength(transaction.data) - 2) / 2
```
Do you mean if `Buffer.byteLength(transaction.data)` is `0`? 

`transaction.data` is a string

hectorgomezv replied:
In that case `transaction.data` would be a falsy value so the ternary operation would return `0`. Maybe there's a corner case if `transaction.data` is `1`, because `dataSize` becomes negative in that case, but reviewing the code it seems it won't cause an error either. I don't have all the context here, I'm just trying to mimic the Rust implementation, do you have any concern about it?

fmrsabino replied:
Ok I see what you mean with `transaction.data` being `0`. In my opinion I think it's harder to read if we treat `""` as being the same as `undefined`.

Probably something like the following what do you think?

```typescript
const dataByteLength = transaction.data
  ? Buffer.byteLength(transaction.data)
  : 0;

const dataSize = dataByteLength >= 2 ? (dataByteLength - 2) / 2 : 0;
```

hectorgomezv replied:
Yes, good catch. I changed it in [912132b](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/912132beed905de946a60c305496cd05313a51ba), thanks!

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,592 @@
+    transaction: MultisigTransaction,
+    safe: Safe,
+  ): Promise<TransactionInfo> {
+    const value = Number(transaction?.value) || 0;
```
The transaction service returns a string representing the number. Do you mean changing it on the Transaction Service side, or maybe coercing it before reaching this point?

fmrsabino replied:
If the transactions service returns it as a `String` but we convert it to `value` we should probably review if we should coerce it.

For now I think this is ok given that we are only directly using it here.

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,592 @@
+    return this.mapCustomTransaction(transaction, value, dataSize, chainId); // TODO: default case
+  }
+
+  private filterAddressInfo(addressInfo: AddressInfo): AddressInfo {
+    return {
+      ...{ value: addressInfo.value },
+      ...(addressInfo.name !== '' && { name: addressInfo.name }),
+      ...(addressInfo.logoUri && { logoUri: addressInfo.logoUri }),
+    };
+  }
```
This is intended to remove `null` values from the returned `AddressInfo`. I mean, the Rust implementation removes properties holding a `null` value so at the end if `name` or `logoUri` are `null`, it returns an object just holding the `value` property. This is intended to mimic that behavior in JS, maybe there's a better/clearer way to do it. I'm open to suggestions 🙂

fmrsabino replied:
Can't this be solved on the entity side with Swagger? I think this is quite verbose and on the CGW it's solved on the entity side – if we do it like this then we need to always apply this logic everywhere else.

Better yet, I'd allow `null`. There should be no distinction in the spec between `null` values and non-present values

hectorgomezv replied:
I do agree, I think allowing `null` shouldn't break anything. I'll do it that way 👍🏻

hectorgomezv replied:
Changed in [4827c02](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/4827c02120e9672707846c48d0552b30dd1952cb)

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,592 @@
+    }
+  }
+
+  private checkSenderOrReceiver(transaction: MultisigTransaction): boolean {
```
Changed in [c42b4c3](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/c42b4c3a1e16fe4d11c566fcd7d75d553c157e34)

fmrsabino commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,599 @@
+    switch (dataDecoded.method) {
+      case MultisigTransactionMapper.SET_FALLBACK_HANDLER:
+        return <SetFallbackHandler>{
+          type: 'SET_FALLBACK_HANDLER',
```
I think we should set the `type` as a property in the class itself. Allowing `SetFallbackHandler` + separately setting the `label` would involve more work.

hectorgomezv replied:
Changed in [c464a50](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/c464a504861327618faa18ddc5c0662aace9d5a0)

hectorgomezv replied:
Explicit constructors were introduced in [c464a50](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/c464a504861327618faa18ddc5c0662aace9d5a0)

fmrsabino commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,599 @@
+    };
+  }
+
+  private async mapSettingsInfo(
+    chainId: string,
+    transaction: MultisigTransaction,
+    safe: Safe,
+  ): Promise<SettingsInfo | undefined> {
```
This function doesn't have a lot of tests to check if the decoding works as expected 😬

hectorgomezv replied:
Tests added in [c9783bc](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/c9783bcef32930fa60fd6c12e8ef339fbdc46b07)

fmrsabino reviewed (changes requested):
I think we are lacking some tests and I've also suggested how we can rely less on setting the types on the route entities.

Let me know if you want to tackle this in a different way.

My suggestion is to Unit test the mapper. There are a lot of conditions that can be cumbersome to write. So I'd isolate the mapper and test the different conditions there.

hectorgomezv commented:
> I think we are lacking some tests and I've also suggested how we can rely less on setting the types on the route entities.
> 
> Let me know if you want to tackle this in a different way.
> 
> My suggestion is to Unit test the mapper. There are a lot of conditions that can be cumbersome to write. So I'd isolate the mapper and test the different conditions there.

Yes, I agree, I think the best way to test all the conditions is to unit-test the mapper. I'll do it 👍🏻

I'm converting the PR to draft in the meanwhile. I'll be addressing your comments as well.

fmrsabino commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,592 @@
+      return {
+        ...executionInfo,
+        missingSigners: this.getMissingSigners(transaction, safe, txStatus),
+      };
```
What I mean is that every time you need to return an `ExecutionInfo` you need to use the spread operator for it to be "correct" right?

I think it's preferable to return the actual instance of `ExecutionInfo` than an implementation (which can be extended of `ExecutionInfo`).

If in order to instantiate a `ExecutionInfo` we need to think about what to include or not include then I think a constructor solution is more simple and also more elegant (easier to read compared to the spread operator).

fmrsabino replied:
Example:

Given the following route entities:

```typescript
export abstract class ExecutionInfo {
  @ApiProperty()
  type: string;

  protected constructor(type: string) {
    this.type = type;
  }
}

export class MultisigExecutionInfo extends ExecutionInfo {
  @ApiProperty()
  nonce: number;
  @ApiProperty()
  confirmationsRequired: number;
  @ApiProperty()
  confirmationsSubmitted: number;
  @ApiPropertyOptional()
  missingSigners: AddressInfo[] | null;

  constructor(
    nonce: number,
    confirmationsRequired: number,
    confirmationsSubmitted: number,
    missingSigners: AddressInfo[] | null,
  ) {
    super('MULTISIG');
    this.nonce = nonce;
    this.confirmationsRequired = confirmationsRequired;
    this.confirmationsSubmitted = confirmationsSubmitted;
    this.missingSigners = missingSigners;
  }
}
```

You can do the following

```typescript
private mapExecutionInfo(
    transaction: MultisigTransaction,
    safe: Safe,
    txStatus: string,
  ): ExecutionInfo {
    const missingSigners =
      txStatus === 'AWAITING_CONFIRMATIONS'
        ? this.getMissingSigners(transaction, safe)
        : null;

    return new MultisigExecutionInfo(
      transaction.nonce,
      this.getConfirmationsRequired(transaction, safe),
      this.getConfirmationsCount(transaction),
      missingSigners,
    );
  }
```

This accomplishes the following:
- You no longer need to specify a `type`. This could result in issues if other strings are applied but we want to enforce a specific type and not any `string`.
- By no longer relying on the spread operator we moved the logic on how to build this entity to the entity itself – which in my opinion avoids mistakes elsewhere.

fmrsabino replied:
Alternatively you could set the type to `type: "MULTISIG"`.

However when Module transactions are introduced you'd need to have `type: "MULTISIG" | "MODULE"`. But the schema for both is quite different (module transactions only have one property: `address`.

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,599 @@
+      id: `multisig_${transaction.safe}_${transaction.safeTxHash}`,
+      timestamp:
+        transaction?.executionDate?.getTime() ??
+        transaction?.submissionDate?.getTime(),
```
Yes, `executionDate` should be always available according to the Transaction Service specification. Maybe the fallback was a legacy thing? I've changed it in [1f6827a](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/1f6827af4f3e7f07a085d74cc91686ad41a0385b)

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,592 @@
+    txStatus: string,
+  ): ExecutionInfo {
+    const executionInfo = {
+      type: 'MULTISIG',
```
Changed in [32a999f](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/32a999fbc2d03262535493692ec03f34f8ff83ff) 👍🏻

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,599 @@
+  private mapTransactionStatus(
+    transaction: MultisigTransaction,
+    safe: Safe,
+  ): TransactionStatus {
```
Tests added in [c9783bc](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/c9783bcef32930fa60fd6c12e8ef339fbdc46b07)

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,599 @@
+      confirmationsSubmitted: this.getConfirmationsCount(transaction),
+    };
+
+    if (txStatus === 'AWAITING_CONFIRMATIONS') {
```
Tests added in [c9783bc](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/c9783bcef32930fa60fd6c12e8ef339fbdc46b07)

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,599 @@
+  }
+
+  private getConfirmationsCount(transaction: MultisigTransaction): number {
+    return transaction.confirmations?.length || 0;
```
Covered in [9b9d1d7](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/9b9d1d7ef7f597aa9f44d919320a43efde273b59)

hectorgomezv commented on src/routes/transactions/mappers/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,599 @@
+    return transaction.confirmationsRequired ?? safe.threshold; // TODO: review if this fallback still required
+  }
+
+  private getMissingSigners(
```
Covered in [9b9d1d7](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/9b9d1d7ef7f597aa9f44d919320a43efde273b59) also

fmrsabino commented on src/routes/transactions/entities/transaction.entity.ts:
```diff
@@ -0,0 +1,35 @@
+  timestamp: number;
+  @ApiProperty()
+  txStatus: string;
+  @ApiProperty()
+  txInfo: TransactionInfo;
```
`@nestjs/swagger` requires a little bit more info here to generate the possible schemas for `TransactionInfo`

You can do something like:

```typescript
@ApiProperty({
  oneOf: [
    { $ref: getSchemaPath(CustomTransactionInfo) },
    { $ref: getSchemaPath(SettingsChangeTransaction) },
    { $ref: getSchemaPath(TransferTransactionInfo) },
  ],
})
txInfo: TransactionInfo;
```

The Transaction class however needs to include these models too:

```typescript
@ApiExtraModels(
  CustomTransactionInfo,
  SettingsChangeTransaction,
  TransferTransactionInfo,
)
export class Transaction {
  ...
}
```

This way the correct schemas will be available via the Swagger interface. The same can be done for `ExecutionInfo` in this class.

If you prefer to tackle this in a separate PR it's also ok 👍

hectorgomezv replied:
You're right, good catch! Included here [d985d6c](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/d985d6cb5043fb4c54881c2cdab35c7007f20306). Thanks for the suggestion!

fmrsabino commented on src/routes/transactions/mappers/multisig-transactions/transaction-info/custom-transaction.mapper.ts:
```diff
@@ -0,0 +1,75 @@
+
+  private getActionCount(transaction: MultisigTransaction): number | null {
+    const { dataDecoded } = transaction;
+    if (dataDecoded?.method === 'multiSend') {
```
nitpick: extract `'multiSend'` (static)

hectorgomezv replied:
Changed in [d985d6c](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/d985d6cb5043fb4c54881c2cdab35c7007f20306)

fmrsabino commented on src/routes/transactions/mappers/multisig-transactions/transaction-info/custom-transaction.mapper.ts:
```diff
@@ -0,0 +1,75 @@
+    const { dataDecoded } = transaction;
+    if (dataDecoded?.method === 'multiSend') {
+      const parameter = dataDecoded.parameters?.find(
+        (parameter) => parameter.name === 'transactions',
```
nitpick: extract `'transactions'` (static)

hectorgomezv replied:
Changed in [d985d6c](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/d985d6cb5043fb4c54881c2cdab35c7007f20306)

fmrsabino commented on src/routes/transactions/mappers/multisig-transactions/transaction-info/custom-transaction.mapper.ts:
```diff
@@ -0,0 +1,75 @@
+      (!gasToken || gasToken === CustomTransactionMapper.NULL_ADDRESS) &&
+      (!refundReceiver ||
+        refundReceiver === CustomTransactionMapper.NULL_ADDRESS) &&
+      (!safeTxGas || safeTxGas === 60)
```
Is `60` a typo?

hectorgomezv replied:
Oh yes, thanks!

Fixed in [d985d6c](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/d985d6cb5043fb4c54881c2cdab35c7007f20306)

fmrsabino commented on src/routes/transactions/mappers/multisig-transactions/transaction-info/erc721-transfer.mapper.ts:
```diff
@@ -0,0 +1,62 @@
+
+@Injectable()
+export class Erc721TransferMapper {
+  private static readonly NULL_ADDRESS =
+    '0x0000000000000000000000000000000000000000';
```
We should extract this. It's already being used in a couple places.

hectorgomezv replied:
Included in [d985d6c](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/d985d6cb5043fb4c54881c2cdab35c7007f20306) 👍🏻

fmrsabino commented on src/routes/transactions/mappers/multisig-transactions/multisig-transaction-execution-info.mapper.ts:
```diff
@@ -0,0 +1,40 @@
+  mapExecutionInfo(
+    transaction: MultisigTransaction,
+    safe: Safe,
+    txStatus: string,
```
Why not pass the mapped `txStatus`? Then we could compare against an enum and not against a `string`

hectorgomezv replied:
Agreed, changed in [d985d6c](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/d985d6cb5043fb4c54881c2cdab35c7007f20306)

fmrsabino commented on src/routes/transactions/entities/multisig-execution-info.entity.ts:
```diff
@@ -0,0 +1,27 @@
+  confirmationsRequired: number;
+  @ApiProperty()
+  confirmationsSubmitted: number;
+  @ApiPropertyOptional({ type: Array<AddressInfo>, nullable: true })
```
I don't think the generic can be applied here (it's expecting an expression) 🤔

hectorgomezv replied:
True, fixed in [d985d6c](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/d985d6cb5043fb4c54881c2cdab35c7007f20306)

fmrsabino commented on src/routes/transactions/mappers/multisig-transactions/multisig-transaction.mapper.ts:
```diff
@@ -0,0 +1,43 @@
+      transaction,
+      safe,
+    );
+    const executionInfo = await this.executionInfoMapper.mapExecutionInfo(
```
`await` shouldn't be required here.

hectorgomezv replied:
Changed in [d985d6c](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/d985d6cb5043fb4c54881c2cdab35c7007f20306)

fmrsabino reviewed (changes requested):
I think the new changes are a big improvement!

I have some additional comments but all of them should be a small thing.

hectorgomezv commented on src/routes/common/entities/address-info.entity.ts:
```diff
@@ -3,14 +3,14 @@ import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
+  @ApiPropertyOptional({ type: String, nullable: true })
+  readonly logoUri: string | null;
 
   constructor(value: string, name?: string, logoUri?: string) {
```
Yes, I thought the same, I think we can tackle that in a separate PR 👍

fmrsabino reviewed (dismissed):
🚀 🚀 🚀

hectorgomezv commented:
> 🚀 🚀 🚀

I just merged your ChainBuilder change in a test class, now it should be ready for approving again, sorry for the inconveniences 🙂
